### PR TITLE
fix: preserve binary file content as base64

### LINKED ISSUE #648: get_file_contents corrupts binary files (PDF, images, archives): Base64 content is force-decoded to UTF-8
## Summary

`get_file_contents` irreversibly corrupts every binary file in a repository (PDF, PNG/JPG, ZIP, XLSX/DOCX, fonts, compiled artifacts). The GitLab API returns file content as Base64, but `getFileContents()` force-decodes it into a UTF-8 **string**. Any byte sequence that is not valid UTF-8 is replaced with `U+FFFD` (`EF BF BD`), so the original bytes are permanently lost before the content ever reaches the MCP client.

Reported originally as "fetching a PDF loses data" — the file is not merely unusable, the byte stream is destroyed and cannot be reconstructed client-side.

## Affected versions

- Confirmed on `@zereight/mcp-gitlab@2.1.30`
- Still present on `main` (`926d42c`) / `2.1.46` — the code is unchanged

## Root cause

[`index.ts#L2054-L2058`](https://github.com/zereight/gitlab-mcp/blob/926d42c8780cb9ec0cf3b5e187575bc3db205ff0/index.ts#L2054-L2058)

```ts
// Decode Base64-encoded file content to UTF-8
if (!Array.isArray(parsedData) && parsedData.content) {
  parsedData.content = Buffer.from(parsedData.content, "base64").toString("utf8");
  parsedData.encoding = "utf8";
}
```

This conversion is **unconditional** — there is no binary detection and no way to opt out.

Note that `GITLAB_REPO_FILE_ENCODING` / `--repo-file-encoding` does **not** help here. It is only referenced in the write paths (`encodeRepoFilePayloadContent()` at L4783, plus L4821 and L4907); `getFileContents()` never consults it. So a user hitting this has no configuration-level workaround.

## Reproduction

No GitLab instance required — this reproduces the exact transformation the server applies:

```js
const header = Buffer.from('%PDF-1.7\n', 'latin1');
const binaryStream = Buffer.from([
  0x25,0xC7,0xEC,0x8F,0xA2,0x0A,
  0x78,0x9C,0xED,0x9A,0x4B,0x6F,0xDB,0x30,
  0x0C,0xC7,0xEF,0xFE,0x8A,0x82,0x9E,0x0A,
  0xFF,0xD8,0xFF,0xE0,0x00,0x10,0x4A,0x46
]);
const footer = Buffer.from('\n%%EOF\n', 'latin1');
const pdf = Buffer.concat([header, binaryStream, footer]);

const apiContent = pdf.toString('base64');          // what GitLab returns

// index.ts:2057
const afterMcp = Buffer.from(apiContent, 'base64').toString('utf8');
const restored = Buffer.from(afterMcp, 'utf8');     // best a client can do

console.log('original bytes :', pdf.length);                       // 46
console.log('restored bytes :', restored.length);                  // 73
console.log('identical      :', pdf.equals(restored));             // false
console.log('U+FFFD count   :', (afterMcp.match(/\uFFFD/g)||[]).length); // 14
```

Output:

```
original bytes : 46
restored bytes : 73
identical      : false
U+FFFD count   : 14
```

Byte-level diff:

```
original: 255044462d312e370a 25 c7 ec8fa2 0a 78 9c ed 9a 4b6f db 300c c7 ef fe 8a 82 9e 0a ffd8 ffe0 00104a460a2525454f460a
restored: 255044462d312e370a 25 efbfbd ec8fa2 0a 78 efbfbd efbfbd 4b6f efbfbd 300c efbfbd efbfbd efbfbd efbfbd efbfbd efbfbd 0a efbfbd efbfbd efbfbd efbfbd 00104a460a2525454f460a
```

Corruption is partial and unpredictable: `ec8fa2` survives because it happens to form a valid UTF-8 sequence, while `c7`, `9c`, `ed`, `ff` each collapse into `efbfbd`. This is why the symptom presents as "the PDF lost some data" rather than a clean failure.

## Expected behaviour

Binary content should reach the client losslessly. A backward-compatible fix is to only decode when the round-trip is lossless, and otherwise pass the Base64 through with `encoding` left as `base64` so the client can decode it itself:

```ts
if (!Array.isArray(parsedData) && parsedData.content) {
  const buf = Buffer.from(parsedData.content, "base64");
  const asText = buf.toString("utf8");
  if (Buffer.from(asText, "utf8").equals(buf)) {
    parsedData.content = asText;      // text file — unchanged behaviour
    parsedData.encoding = "utf8";
  } else {
    parsedData.encoding = "base64";   // binary — preserve bytes
  }
}
```

This keeps the current behaviour for all valid-UTF-8 files, so existing text workflows are unaffected.

## Related: the write path has the same weakness

`encodeRepoFilePayloadContent()` does `Buffer.from(content)` with no explicit encoding, which defaults to UTF-8. Uploading binary content through `create_or_update_file` therefore corrupts it in the same way. Worth addressing together, since a round-trip (read then write) currently destroys data twice.

### PR DESCRIPTION
**Pull request by quseijuro-design**
## Summary

- preserve the original Base64 payload and `base64` encoding for non-UTF-8 repository files
- keep the existing decoded text response and `utf8` encoding for valid UTF-8 files
- add MCP-level regression coverage for both binary and text content

## Root cause

`getFileContents` unconditionally decoded every GitLab Base64 payload with `Buffer.toString("utf8")`. Invalid UTF-8 bytes were replaced with the Unicode replacement character, so binary content could not be reconstructed from the response.

The updated read path only returns decoded text when converting the bytes to UTF-8 and back is lossless. Otherwise, it leaves GitLab's original Base64 response unchanged.

## Impact

Clients can now retrieve non-UTF-8 files without data corruption, while text-file behavior remains unchanged. This does not change the write path or public tool arguments.

## Validation

- `npm.cmd run build`
- `node --import tsx/esm --test --experimental-test-isolation=none --test-concurrency=1 test/test-get-file-contents.ts test/streamable-http-concurrent-session.test.ts` (4 tests passed)
- `npx.cmd prettier --check test/test-get-file-contents.ts`
- `npm.cmd run check:runtime-deps`
- `npx.cmd tsx test/oauth-tests.ts`

Closes #648

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -2052,10 +2052,15 @@ async function getFileContents(
...
   if (!Array.isArray(parsedData) && parsedData.content) {
-    parsedData.content = Buffer.from(parsedData.content, "base64").toString("utf8");
-    parsedData.encoding = "utf8";
+    const decodedContent = Buffer.from(parsedData.content, "base64");
+    const utf8Content = decodedContent.toString("utf8");
+
+    if (Buffer.from(utf8Content, "utf8").equals(decodedContent)) {
+      parsedData.content = utf8Content;
+      parsedData.encoding = "utf8";
+    }
```
[the remark this answers is not in the document]
Comment (quseijuro-design) [reply index.ts:2063]: Fixed in bbd9402. The guard now checks that `content` is a string, so an empty Base64 payload follows the same lossless UTF-8 path and is returned with `encoding: utf8`. I also added an MCP-level regression test for an empty file. The focused file-content tests pass (3/3), and the related integration set passes (5/5).

**Review by zereight — APPROVED** [review]
LGTM. Verified binary Base64 preservation, UTF-8 text decoding, and empty file handling with mock test suite.
